### PR TITLE
Feature: PerfEvent support.

### PR DESCRIPTION
**Pull request by Skrinikov**
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this
  change?)
   * Adds the ability to attach perf events via `attach_perf_event`
* what changes does this pull request make?
   * Adds PerfEvents to`bcc::core`
   * Adds an example which uses perf events so to demonstrate how it works.
* are there any non-obvious implications of these changes? (does it break compatibility with previous
  versions, etc)
1. Function signature is a bit heavy. (Python code has a bunch of default values)
2. Not sure whether the enums `PerfType`, `PerfHardwareConfig` and `PerfSoftwareConfig` are added correctly. 
3. Maybe have a better error message handling (You can't run hardware perf on a VM)
4. If no `cpu` is provided we use `cpuonline::get` per call to `attach_perf_event` (Some way to cache result)

### CODE REVIEW

File: src/core/perf_event/mod.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#[cfg(any(
```
Comment (brayniac) [inline src/core/perf_event/mod.rs:1]: I don't think you need the cfg block here, we seem to always include this particular module

File: src/core/perf_event/mod.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    feature = "v0_11_0",
+    feature = "v0_12_0",
+    feature = "v0_13_0",
+    feature = "v0_14_0",
+    feature = "v0_15_0",
+    not(feature = "specific"),
+))]
+mod v0_4_0;
+
+#[cfg(any(
```
Comment (brayniac) [inline src/core/perf_event/mod.rs:19]: Same here

File: src/perf.rs [hunk 3]
```diff
@@ -10,6 +10,50 @@ use crate::table::Table;
 use crate::types::*;
 use crate::BccError;
 
+// TODO:
+//    - The enums below should probably generated in bcc-sys
+//    - Find a way to map generated to
+
+#[allow(dead_code)]
+pub enum PerfType {
```
Comment (brayniac) [inline src/perf.rs:18]: This enum doesn't look like it covers the other types, we definitely want to support the other types of perf events here too. We have current use cases for the hw cache type and the raw type is likely to be useful in the near term

**Comment by brayniac** [comment]
Overall, looks good. I'd like to have an example that uses some software event that we can run in CI. Maybe cpu migrations or  context switches would make a good smoke test.

For the function signature, I was thinking of ways we can start moving to a builder pattern in this library. Does not need to be addressed in this PR

File: examples/contextswitch.c [hunk 4]
```diff
@@ -0,0 +1,42 @@
+
```
Comment (brayniac) [inline examples/contextswitch.c:1]: Drop the leading newline before the license text

File: examples/contextswitch.rs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+use bcc::core::BPF;
+use bcc::perf::{PerfSoftwareConfig, PerfType};
+use bcc::BccError;
+use clap::{App, Arg};
+
+use core::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::{mem, ptr, thread, time};
+
+// Both consants are arbitrary
```
Comment (brayniac) [inline examples/contextswitch.rs:14]: what are the units?

Comment (Skrinikov) [reply examples/contextswitch.rs:14]: FREQ = Hertz
DURATION = seconds

File: examples/contextswitch.rs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        .arg(
+            Arg::with_name("pid")
+                .long("pid")
+                .help("Only track this pid")
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("tgid")
+                .long("tgid")
+                .help("Only track this thread")
```
Comment (brayniac) [inline examples/contextswitch.rs:56]: tgid is for a thread group. Confused by the help for this arg

File: examples/contextswitch.rs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    let mut sample_frequency: Option<u64> = matches
+        .value_of("sample_frequency")
+        .map(|v| v.parse().expect("Invalid sample frequency"));
+
+    let sample_period: Option<u64> = matches
+        .value_of("sample_period")
+        .map(|v| v.parse().expect("Invalid sample period"));
+
+    if !sample_frequency.is_some() && !sample_period.is_some() {
```
Comment (brayniac) [inline examples/contextswitch.rs:75]: `is_none()` might be nicer

File: examples/contextswitch.rs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        sample_period,
+        sample_frequency,
+        None,
+        None,
+        None,
+    )?;
+
+    println!("Running for {} seconds", duration);
+
+    let mut durr = 0;
```
Comment (brayniac) [inline examples/contextswitch.rs:111]: maybe elapsed is a better name?

File: examples/contextswitch.rs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+            println!("{:<-8} {:>12}", key, value);
+        }
+    }
+
+    Ok(())
+}
+
+fn parse_u32(x: Vec<u8>) -> u32 {
+    let mut v = [0_u8; 4];
+    for i in 0..4 {
```
Comment (brayniac) [inline examples/contextswitch.rs:147]: Take a look at the changes I made in: https://github.com/rust-bpf/bpf-tools/blob/master/src/hardirqs/main.rs#L163

File: examples/contextswitch.rs (hunk 5 above)
Comment (brayniac) [reply examples/contextswitch.rs:14]: Can you add a comment next to both of the chosen defaults? =)

**Comment by brayniac** [comment]
There are a couple other lints that you might want to address: https://github.com/rust-bpf/rust-bcc/pull/106/checks?check_run_id=920461866#step:4:40

**Comment by Skrinikov** [comment]
> There are a couple other lints that you might want to address: https://github.com/rust-bpf/rust-bcc/pull/106/checks?check_run_id=920461866#step:4:40

Changed everything besides the fn signature max argument count for now.

not sure why format fails right now.

File: examples/contextswitch.rs [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+            Arg::with_name("duration")
+                .long("duration")
+                .short("d")
+                .help("How long to run this trace for (in seconds)")
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("pid")
+                .long("pid")
+                .help("Only track this TID")
```
Comment (brayniac) [inline examples/contextswitch.rs:50]: typo? Believe this should be `PID`

File: examples/llcstat.rs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    let duration: u64 = matches
+        .value_of("duration")
+        .map(|v| v.parse().expect("Invalid duration"))
+        .unwrap_or(DEFAULT_DURATION);
+
+    let code = include_str!("llcstat.c").to_string();
+    let mut bpf = BPF::new(&code)?;
+    bpf.attach_perf_event(
+        "on_cache_miss",
+        PerfType::Hardware as u32,
```
Comment (brayniac) [inline examples/llcstat.rs:57]: I don't like having to do `as u32` at the call site. Consider making the function take generic types which have a type bound `u32: From<T>`

File: src/perf.rs [hunk 12]
```diff
@@ -10,6 +10,68 @@ use crate::table::Table;
 use crate::types::*;
 use crate::BccError;
 
+#[allow(dead_code)]
```
Comment (brayniac) [inline src/perf.rs:13]: I'm surprised this is needed since the enum is public...

File: examples/llcstat.rs (hunk 11 above)
Comment (brayniac) [reply examples/llcstat.rs:57]: Deferring this to a future refactor. I need to think about the design more carefully

**Pull request merged**